### ISSUE TITLE
Doesn't respect default-extensions from .cabal file

### ISSUE THREAD

**Issue by lylek** [facebookincubator/retrie#18]
I got an error indicating that I should enable an extension:

```
retrie: user error ((RealSrcSpan SrcSpanOneLine "/Users/lylek/blah/src/Foo.hs" 91 34 38,"Illegal lambda-case (use -XLambdaCase)"))
```

But I have no way of enabling extensions via retrie. It would be nice if it could read and use the `default-extensions` values from the `.cabal` file. Or at least have a way of passing a list of extensions as an option to `retrie`.

**Comment by pepeiborra** [comment]
The retrie integration in HLS (Haskell language server) takes care of this

**Comment by lylek** [comment]
OK, thanks. I haven't tried HLS yet because I'm working with GHC 8.2, which isn't supported by HLS. But I'll be switching to GHC 8.8 relatively soon.

In this particular case, I was trying to do an adhoc rewrite on a particular file. Is that easily doable with HLS?

**Comment by anka-213** [comment]
I was thinking the same thing. The HLS integration is somewhat limited in supported features, so it would be nice to have cabal support in the command line tool as well.

**Comment by xich** [comment]
I'm not sure what is required to parse and extract extensions from a .cabal file. Presumably the cabal library can do this?

If someone were to add it, the place to do it would be in `resolveOptions`, here:

https://github.com/facebookincubator/retrie/blob/master/Retrie/Options.hs#L322

(because `addLocalFixities` is the first place retrie parses something)

**Comment by lylek** [comment]
One tricky bit is that there can be a `default-extensions` for each library/executable. I'm not sure how one would resolve the right one, but presumably the cabal library can figure that out. Would you have to specify a build target to retrie?

**Comment by ilyakooo0** [comment]
So basically this means if you use _any_ language extensions which modify the syntax (LambdaCase, PatternSynonims, etc) there is no way at all to use this utility with the project. I think this is quite a serious limitation.

**Comment by pepeiborra** [comment]
Given that HLS already takes care of setting up your project via hie-bios, perhaps the simplest option would be to embed retrie in HLS as a new command. HLS plugins cannot add new commands yet, but we can fix that. 

```
haskell-language-server retrie --adhoc ...
```

Alternatively, we could hook up retrie to hie-bios for the GHC session setup.

**Comment by ilyakooo0** [comment]
As a simple solution: supplying an extension list via the CLI would also be suitable.

**Comment by xich** [comment]
@pepeiborra I did some initial digging around today. It looks like HLS has a `HscEnv` available when it calls retrie via the plugin. So my thought was to make retrie's run functions accept an `HscEnv` and redo the parsing functions to get the dynflags from it, rather than whatever bootstrapping ghc-exactprint does. Then when retrie is invoked by HLS, it uses the `HscEnv` that HLS already has available. When retrie is invoked outside HLS, I'll use hie-bios directly to create an HscEnv. I think this can be built upon to solve #10 as well.

Does that sound like a reasonable approach?

Regarding using hie-bios directly, `loadImplicitCradle`/`getCompilerOptions`/`initSession` seem to be the things that do what I want... but it seems like I need to call them on each target file, rather than once per project root (retrie's `targetDir`). Is that correct? Since `initSession` sets up a temp directory for interface caching, it seems like I would really want to just set this up once, rather than once per target file. Especially as I look to applying this to #10.

**Comment by xich** [comment]
Also, I haven't looked at how HLS sets up hie-bios stuff yet. If there is a lot of hard won magic there that I would have to duplicate when calling hie-bios directly, maybe I should just make the retrie binary depend on HLS so I can reuse it instead of hie-bios directly?

**Comment by pepeiborra** [comment]
Yeah exactly, if retrie main entry point can accept an `HscEnv`, I can easily integrate it as an HLS command line subcommand.

Concretely, I had in mind defining a custom CLI command. Custom commands get an `IdeState` which allows them to use `runAction` and have access to all the build rules.

https://github.com/haskell/haskell-language-server/blob/861c8bf4b48434dd1ad9b386a3fcd44bbca04f23/ghcide/src/Development/IDE/Main.hs#L95-L101

Currently the set of custom commands is fixed, but I think it should be easy to allow plugins to declare new commands that can be used from the CLI. Then hls-retrie-plugin could define a new CLI command that exposes all of the retrie functionality.

I'm not sure how much code duplication it would be to drive hie-bios directly. HLS takes care of this and also hooks up with implicit-hie so that a cradle is not needed in most cases.
